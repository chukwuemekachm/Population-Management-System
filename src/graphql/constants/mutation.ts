import { gql } from 'apollo-server-express';

export const SIGNUP_MUTATION = gql`
  mutation signup($user: SignupInput) {
    signup(user: $user) {
      user {
        email
        firstName
        lastName
      }
    }
  }
`;

export const LOGIN_MUTATION = gql`
  mutation login($user: LoginInput) {
    login(user: $user) {
      user {
        firstName
        lastName
      }
    }
  }
`;
