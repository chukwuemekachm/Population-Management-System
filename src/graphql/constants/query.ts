import { gql } from 'apollo-server-express';

export const GET_LOCATIONS_QUERY = gql`
  query getLocations($filter: LocationFilterInput) {
    getLocations(filter: $filter) {
      locationName
    }
  }
`;
