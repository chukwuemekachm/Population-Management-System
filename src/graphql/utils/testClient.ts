import { ApolloServer } from 'apollo-server-express';
import { createTestClient } from 'apollo-server-testing';

import typeDefs from '../schema';
import resolvers from '../resolvers';
import prismaTestMockClient from '../__mocks__/prisma';
import { User } from '../types/types';

export const constructGraphqlTestClient = (user?: Partial<User>) => {
  const testServer = new ApolloServer({
    typeDefs,
    resolvers,
    context: ({ req }) => ({
      request: req,
      prisma: prismaTestMockClient,
      user,
    }),
  });

  return createTestClient(testServer);
};
