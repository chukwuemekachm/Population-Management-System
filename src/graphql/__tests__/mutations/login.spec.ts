import prismaTestMockClient from '../../__mocks__/prisma';
import { LOGIN_MUTATION, SIGNUP_MUTATION } from '../../constants';
import { userInputs } from '../../fixtures/userFixtures';
import { constructGraphqlTestClient } from '../../utils/testClient';

const { mutate } = constructGraphqlTestClient();

describe('login mutation', () => {
  beforeAll(async () => {
    await prismaTestMockClient.deleteManyLocations();
    await prismaTestMockClient.deleteManyUsers();
    const payload = {
      variables: {
        user: userInputs[1],
      },
    };
    await mutate({
      mutation: SIGNUP_MUTATION,
      ...payload,
    });
  });
  afterAll(async () => {
    await prismaTestMockClient.deleteManyLocations();
    await prismaTestMockClient.deleteManyUsers();
  });

  test('should return authenticated user profile when credentials are valid', async () => {
    const payload = {
      variables: {
        user: {
          email: userInputs[1].email,
          password: userInputs[1].password,
        },
      },
    };
    const { errors, data } = await mutate({
      mutation: LOGIN_MUTATION,
      ...payload,
    });
    expect(errors).toBeUndefined();
    expect(data).toBeDefined();
    expect(data).toMatchSnapshot();
  });

  test('should return validation errors when inputs are not provided', async () => {
    const payload = {
      variables: {},
    };
    const { errors, data } = await mutate({
      mutation: LOGIN_MUTATION,
      ...payload,
    });
    expect(errors).toBeDefined();
    expect(data).toBeNull();
    expect(errors).toMatchSnapshot();
  });

  test('should return an unauthenticated error response when the user does not exist', async () => {
    const payload = {
      variables: {
        user: {
          email: userInputs[2].email,
          password: userInputs[2].password,
        },
      },
    };
    const { errors, data } = await mutate({
      mutation: LOGIN_MUTATION,
      ...payload,
    });
    expect(errors).toBeDefined();
    expect(data).toBeNull();
    expect(errors).toMatchSnapshot();
  });

  test('should return an unauthenticated error response when the password does not match', async () => {
    const payload = {
      variables: {
        user: {
          email: userInputs[1].email,
          password: 'Mismatch',
        },
      },
    };
    const { errors, data } = await mutate({
      mutation: LOGIN_MUTATION,
      ...payload,
    });
    expect(errors).toBeDefined();
    expect(data).toBeNull();
    expect(errors).toMatchSnapshot();
  });
});
