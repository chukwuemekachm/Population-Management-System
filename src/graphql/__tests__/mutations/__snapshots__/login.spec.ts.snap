// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login mutation should return an unauthenticated error response when the password does not match 1`] = `
Array [
  [GraphQLError: Invaild credentials],
]
`;

exports[`login mutation should return an unauthenticated error response when the user does not exist 1`] = `
Array [
  [GraphQLError: Invaild credentials],
]
`;

exports[`login mutation should return authenticated user profile when credentials are valid 1`] = `
Object {
  "login": Object {
    "user": Object {
      "firstName": "Test",
      "lastName": "User",
    },
  },
}
`;

exports[`login mutation should return validation errors when inputs are not provided 1`] = `
Array [
  [GraphQLError: Some field(s) are failing validation],
]
`;
